### DISCUSSION TITLE
Copying test file from docker container to GitHub repository

### DISCUSSION THREAD

**Discussion by tdalvi** [community/community#26791] [Actions]
<p>Hi. Apologies if this has already been answered. I have a github action that is setup to run a selenium test in a docker container. The test creates a TestResult.html file on the docker container. I am binding a volume that contains the test dll to “/test” folder on the docker container. The TestResult.html is also created in the “/test” folder on the docker container. When I execute this manually from my local machine, the volume that contains the test dll and the “/test” folder are in sync and the TestResult.html file is saved in the folder on my local machine automatically. However, when I run this through GitHub actions, the test runs successully but nothing is saved in the synced volume on GitHub. Is there a way to copy this result file from the docker container into my github repository? I tried adding the step “docker cp MyContainer:Path/to/file ${{ github.workspace }}/Artifacts” in my action.yml where Artifacts is a folder I have created in my GitHub repository. The step shows as passed but no file is saved in the Artifacts folder.  Thanks.</p>

**Comment by airtower-luna** [comment]
<p>The build results on the VM provided by Actions are separate from your repository, unlike on your local file system. The repository is checked out, the workflow runs, and then the VM (including its files) is discarded.</p>
<p>The best solution to archive your test results is probably to store them as an artifact (<a href="https://help.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts" title="Persisting workflow data using artifacts" rel="nofollow noopener">documentation here</a>). You can then download the artifact for each run.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/tdalvi">@tdalvi</a> ,</p>
<p>When you run the Docker container action in workflow, from the logs for this action execution step, you can see it will run some commands like below:</p>
<pre><code>/usr/bin/docker run
...
-v "/home/runner/work/_temp/_github_home":"/github/home"
-v "/home/runner/work/_temp/_github_workflow":"/github/workflow"
-v "/home/runner/work/myRepo/myRepo":"/github/workspace"
...
</code></pre>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/7/782466de0171a1b802006dcf266e4257ff73ccbf.png" data-download-href="/uploads/short-url/h8PhNfayQSUHQOjcxL3lBO0g8LZ.png?dl=1" title="example.PNG" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/46307996/181099928-a4e9effc-467c-4f5b-bcd2-a2a95880fc6e.png" alt="example.PNG" data-base62-sha1="h8PhNfayQSUHQOjcxL3lBO0g8LZ" width="690" height="47" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/7/782466de0171a1b802006dcf266e4257ff73ccbf_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">example.PNG</span><span class="informations">1523×105 12 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>This will automatically mount volumes on the container and on the runner machine, it can be used to share/sync files between the container and the runner machine.</p>
<p>For example, if you create a " <strong>test</strong>" folder in the " <strong>/github/workspace</strong>" directory on the container, and add the " <strong>TestResult.html</strong>" file into the " <strong>test</strong>" folder. Then these operations will sync to the " <strong>/home/runner/work/myRepo/myRepo</strong>" directory, it will automatically generate " <strong>/home/runner/work/myRepo/myRepo/test/TestResult.html</strong>"</p>
<p>So after executing the Docker container action to generate the the " <strong>TestResult.html</strong>" on the specified directory (I recommend you set to generate this file into one of the automatically mount volumes mentioned above), you can add an extra step in the same job to copy the file to another directory on the runner machine. For example, copy " <strong>/home/runner/work/myRepo/myRepo/test/TestResult.html</strong>" to another directory.</p>
<p>You should understand that this only can work on the same job, if you also want to share the " <strong>TestResult.html</strong>" file between other jobs in the same workflow run, you need to upload it as an <strong>artifact</strong> , just like <a class="mention" href="/airtower-luna">@airtower-luna</a> has mentioned.</p>

**Comment by BrightRan** [comment]
<p><a class="mention" href="/tdalvi">@tdalvi</a> ,</p>
<p>How are things going? Is my above suggestion helpful to you?</p>
<p>If you have any question about this ticket, feel free to contact us.</p>

**Comment by tdalvi** [comment, marked as answer]
<p><a class="mention" href="/brightran">@brightran</a> <a class="mention" href="/airtower-luna">@airtower-luna</a> Thank you both for the suggestions. I managed to solve this issue last Sunday. Here is what I did. After the test run was complete and the test file created on the container, I ran the command “docker cp name of container:/path/to/testfile/TestResults.html .”.  In case it is not obvious, there is a “.” at the end of that statement. This copied my test file from the container to the location “/home/runner/work/myrepo/myrepo”. Then, in the next step I used the upload artifact as <a class="mention" href="/airtower-luna">@airtower-luna</a> suggested. For the “Path”, I entered “./TestResults.html”. The file was uploaded as an artifact.</p>

**Comment by kspoo** [comment]
<p><a class="mention" href="/tdalvi">@tdalvi</a> I am doing a similar functionality but running into issues with generating the test file in docker when I run it from GitHub Actions. It works in local and html gets added to test reports folder. How can I make sure I can generate the report and use volumes to copy to the local folder?</p>
<p>Below is what I have in my yaml file.<br>
jobs:<br>
build:</p>
<pre><code>runs-on: ubuntu-latest
services:
    selenium:
        image: selenium/standalone-chrome
        ports:
            - "4444:4444"
steps:
- uses: actions/checkout@v2
- name: Set up Python 3.9
  uses: actions/setup-python@v2
  with:
    python-version: 3.9
- name: Install dependencies
  run: |
    python -m pip install --upgrade pip
    pip install -r requirements.txt
- name: Run Tests
  run: python -m unittest testcase/STS/Sanity.py</code></pre>
